docs(api, docs): port Python API documentation to mkdocs

ecormany opened the pull request:
# Overview

This PR ports the entire Python Protocol API documentation from Sphinx to mkdocs.

## Test Plan and Hands on Testing

[Sandbox](https://sandbox.docs.opentrons.com/what-if-no-sphinx-for-real/python-api/)

## Changelog

The overall process for building the new site was to run LLM conversion of RST to markdown, inspect the results, and iterate until the new mkdocs site accurately reflected all the content in the `edge` sandbox.

The commit log for this PR tells a more complete story of the process for conversion.

### What this PR touches
- 🆕 Adds files for the new `/docs/python-api/` mkdocs project.
  - Markdown source files
  - Copy of image assets
- ✍️ Modifies files in `/docs/` that control our top-level docs build.
  - `/docs/mkdocs.yml` includes the new project.
  - Adds the [mkdocstrings plugin](https://github.com/mkdocstrings/mkdocstrings) for rendering the API reference.
  - Adds a custom griffe plugin for showing API level decorators within the API reference.
  - Adds the [macros plugin](https://github.com/fralau/mkdocs-macros-plugin) for substituting the latest API version string in code snippets and prose.
  - Watches the `/api/src/opentrons/protocol_api/` directory for changes when running `make serve`, so if you're editing docstrings you can see the changes automatically.
- 🔗 Updates links in other mkdocs publications that used to go to absolute `docs.opentrons.com/v2/...` URLs.
- 🔨 Updates Makefile to properly pull in latest docstrings and work around an `mkdocs serve` bug.
- 🚨 Updates _every single public docstring in `/api/src/opentrons/`_ to work with markdown parsing and the new API reference. There should be no code changes except at the whim of `black` formatting.

### What this means for others
- **No more Sphinx-style docstrings.** The branch name says it all. This is for real. If you are an engineer writing docstrings for a new feature, use [Google-style docstrings](https://google.github.io/styleguide/pyguide.html#383-functions-and-methods) and markdown formatting. You will never forget the second layer of backticks again. 🎉 
- **New Python object cross-reference style.** The mkdocstrings plugin creates header IDs for each entry in the API reference, which is the fully qualified Python object name. For example, the header ID for `transfer()` is `#opentrons.protocol_api.InstrumentContext.transfer`. Because we use the [autorefs plugin](https://github.com/mkdocstrings/autorefs) for mkdocs, you can use a reference-style markdown link to _any_ unique header _anywhere_ in our documentation (including in publications "outside of" the API docs). So the link to `transfer()` should look like 
  ```
  [`transfer()`][opentrons.protocol_api.InstrumentContext.transfer]
  ```
  Since you manually control the link text:
  - Include backticks to get monospace text.
  - Include parentheses for method names and exclude them for attribute names.
  - Write out the class name, e.g. `InstrumentContext.transfer()` if it feels relevant.
- **Python API docs get good search.** #finally. No seriously, try it. Type your favorite API feature into the search bar in the sandbox. 

### To do
- [x] Figure out how we will handle redirects.
  - The old URLs are bad (all the files with `new_` prefixes) and I've wanted to get rid of them forever. Moving from `docs.opentrons.com/v2/` to `docs.opentrons.com/python-api/` gives us an opportunity to change these filenames (and this PR does).
  - I think redirects are probably best handled in server config rather than in mkdocs. We don't want to output files to a `site/v2/` directory at all.
- [x] Inform teams that the change is coming.
- [x] **Incorporate Sphinx docs changes made after `docs@2.26`** in a follow-up PR
  - #20200
  - #20027 
  - #20217
  - #20240
- [x] Clean out unused RST files in `/api/docs/v2/`
- [x] Leave behind other `/api/docs/` source for archived sites.
- [x] Leave Sphinx tooling in place in case we ever have to rebuild those sites for some reason.

## Review requests

Thank you all for wading in to review a massive PR.

- Technical content team: Ensure that the converted docs retain all of the same content of the original docs, without errors.
- Infra and QA: `make` sure that my changes to `uv` setup and docs build are cromulent.
- Engineers: Please please please triple-check that I am not messing anything up in code. And let me know if anything about this system seems like a downgrade rather than an upgrade for docs-as-code in the future.

## Risk assessment

Medium. Touches every single Python API docstring. Updates docs build process.

ecormany committed:
partway through conversion

ecormany committed:
full conversion of article pages

ecormany committed:
renamed files

ecormany committed:
mkdocs nav

ecormany committed:
some fixups

ecormany committed:
fix up Labware page

ecormany committed:
some Labware changes

ecormany committed:
fix up through Modules landing page + find/replace link formatting

ecormany committed:
fixes through H-S

ecormany committed:
more module page fixes

ecormany committed:
thermocycler changes

ecormany committed:
add mkdocstrings to docs dependencies

ecormany committed:
moam and some of deck slots

ecormany committed:
rest of deck slots

ecormany committed:
edits to loading pipettes; move files to building-block-commands directory

ecormany committed:
fix up pipette characteristics page

ecormany committed:
update partial tip pickup

ecormany committed:
update volume modes

ecormany committed:
update liquid classes

ecormany committed:
liquid class tables, oh my

ecormany committed:
building block up through touch tip

ecormany committed:
finish building block liquid commands

ecormany committed:
through complex commands

ecormany committed:
edit through robot position

ecormany committed:
reorg RTP files

ecormany committed:
edits to RTP pages

ecormany committed:
edit advanced control page

ecormany committed:
update examples page

ecormany committed:
edit adapting page

ecormany committed:
load editable opentrons package for ref

ecormany committed:
convert ProtocolContext docstrings to google style

ecormany committed:
fix up doc headers and style

ecormany committed:
api ref css improvements

ecormany committed:
config: watch for changes in PAPI docstrings

ecormany committed:
convert InstrumentContext docstrings through transfer()

ecormany committed:
convert remaining InstrumentContext docstrings

ecormany committed:
convert Labware docstrings up to load_liquid

ecormany committed:
convert remaining Labware docstrings + WasteChute and TrashBin

ecormany committed:
convert plate reader and general ModuleContext dostrings

ecormany committed:
convert Stacker docstrings

ecormany committed:
convert H-S docstrings

ecormany committed:
convert remaining module docstrings

ecormany committed:
convert type docstrings

ecormany committed:
convert well and liquid docstrings

ecormany committed:
update execute and simulate docstrings; that's all of them!

ecormany committed:
add title metadata and remove H1s

ecormany committed:
first attempt at griffe extension for version numbers; functions only

ecormany committed:
griffe labels working for properties too (thx Seth)

ecormany committed:
remove print debugging in griffe script

ecormany committed:
convert RobotContext docstrings; hide docstrings throughout ref

ecormany committed:
fix links in API ref

ecormany committed:
split module ref entries into separate pages

ecormany committed:
fix up doc links

ecormany committed:
fix ALL the warnings!

ecormany committed:
remove unused html and rst reference files

ecormany committed:
convert docstrings affected by merging in current edge

ecormany committed:
fix a couple links; update makefile to work around mkdocs serve bug

ecormany committed:
remove future dynamix mix link

ecormany committed:
content fixups for welcome, versioning, tutorial, labware

ecormany committed:
fix up moving labware

ecormany committed:
fix up modules through Absorbance Plate Reader; add Stacker page

ecormany committed:
content fix ups through all modules

ecormany committed:
content fixups through pipette characteristics

ecormany committed:
fix up through all pipette pages

ecormany committed:
content fix ups through building blocks

ecormany committed:
content fix ups for complex commands

ecormany committed:
content fixups for order of operations

ecormany committed:
content fix ups for complex command params

ecormany committed:
content fix ups for robot position; delete Axis Speed Limits § as never added to PE

ecormany committed:
content fix ups for runtime parameters

ecormany committed:
new content and refactor of advanced control into multiple pages

ecormany committed:
fix up examples through loops §

ecormany committed:
content fix ups through all pages

ecormany committed:
sentence case headers

ecormany committed:
add macros plugin for apiLevel replacement

ecormany committed:
fix bad previous conflict resolution; blacken

ecormany committed:
zap some double backticks

ecormany committed:
now where did those come from?

ecormany committed:
add black to format signatures; touch up ProtocolContext docstrings

ecormany committed:
fix up InstrumentContext docstrings

ecormany committed:
fix up RobotContext docstrings

ecormany committed:
fix up Labware docstrings

ecormany committed:
fix up Well and LiquidClass docstrings

ecormany committed:
fix up all remaining docstrings

ecormany committed:
the big one: move v2 -> python-api

ecormany committed:
incoming cross-references; move api-reference -> reference for cleaner URL

ecormany committed:
opentrons dependency wasn't necessary

ecormany committed:
or maybe it was, who can say

ecormany committed:
let's try the makefile instead

ecormany committed:
none more black

ddcc4 commented on api/src/opentrons/protocol_api/_liquid.py:
```diff
@@ -29,7 +29,7 @@ class Liquid:
         display_color: An optional display color for the liquid.
 
-    .. versionadded:: 2.14
+    *New in version 2.14*
```
Hm, I'm assuming mkdocs doesn't support as many of these annotations as sphinx, and that's why you had to change it to English text?

ecormany replied:
Yes, that's right. I think nothing is lost in this change. Syntactically, I could never remember the exact Sphinx incantation for these. Semantically, their only target was to become human-readable messages — they were never collated or cross-referenced or anything like that. So here I made the human-readable message directly. 

It also has allowed for a little more flexibility. We had some awkward messages of the sort _Changed in version 2.x: Removed foo._ and now we can just say _Removed in version 2.x: Foo._

ddcc4 commented on api/src/opentrons/protocol_api/disposal_locations.py:
```diff
@@ -78,7 +78,7 @@ def height(self) -> float:
     """Represents a Flex or OT-2 trash bin.
 
-    See :py:meth:`.ProtocolContext.load_trash_bin`.
+    See [`load_trash_bin()`][opentrons.protocol_api.ProtocolContext.load_trash_bin].
```
Heh, I don't think I'm going to check every line of this PR, but I assume most of these changes are mechanical? 

<img width="583" height="71" alt="Image" src="https://github.com/user-attachments/assets/8f906c89-b43e-48f1-af1c-1665362d2a18" />

ecormany replied:
Yes. This prompt did great work.
> convert this docstring from sphinx-style to google-style. keep all wording exact. replace rst formatting with markdown. replace :ref:, :py:meth:, and :py:obj: crossreferences with markdown autoref style links, for example [`transfer()`][opentrons.protocol_api.InstrumentContext.transfer]. the link text for methods should have trailing parentheses, like transfer(), and the link text for other objects should not. when hard wrapping a paragraph or continuing an argument definition past one line, always indent the second and subsequent lines by one level (four spaces) more than the first line. replace RST versionadded and versionchanged statements with italic text saying "New in version…" or "Changed in version…" convert rst admonitions to Material for markdown style admonitions, for example beginning with "!!! note" and indenting all admonition content by one additional level.

And mkdocs was good at raising build warnings if links went to nonexistent targets. I cleaned those up by hand.

ddcc4 commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -207,62 +208,62 @@ def aspirate(  # noqa: C901
+                until the pipette is full. On API levels at or above 2.16, no liquid
+                will be aspirated.
+            location:
+                Tells the robot where
```
For the other reviewers: Try adding `?w=1` to the Github URL to ignore whitespace changes (so this file would be https://github.com/Opentrons/opentrons/pull/20121/files?w=1#diff-6811e2f4d45b8d159abed031a1b00c50e69ecc5aa7d8e19a674527fb4be3298c ).

But @ecormany: Did you deliberately have to re-wrap these lines for mkdocs? Why is the linebreak so strange-looking here?

ecormany replied:
A lot of necessary rewrapping did happen, although this one is admittedly peculiar.

A lot of our old sphinx docstrings had extra indentation, with definitions for things like `:param my_super_long_cool_parameter_name:` indented by 30+ spaces when exactly 4 would have done.

Mkdocs is very strict on having exactly 4 spaces between `Args:` and the start of each arg definition, and another 4 spaces for any subsequent lines of the definition. Markdown requires that lists _not_ be further indented, while Sphinx requires that they _do_. 

Ultimately, what matters is that these files 1. output correctly and 2. pass `black`. Anything else is bonus.

y3rsh commented:
Thoughts in the for of a PR, felt like the clearest way https://github.com/Opentrons/opentrons/pull/20227

emilyburghardt reviewed (approved):
I took a look through, tested links and checked some text sections back and forth between prod and sandbox. Looks good!! 

One thought I had while scrolling through was that our `versionadded` and `versionchanged` messages, which use a specific syntax in Sphinx, now appear without a punctuation in some cases (for example, if it's just `versionadded` with no message about what was added, this appears without a period at the end). Not a blocker to get this merged, but they stuck out in my brain while I was clicking through. Could be backlog.

ecormany committed:
implement feedback from #20227

jwwojak reviewed (approved):
As this documentation has long been in production, providing meaningful input to 100s of pages in a typical docs review may not apply here. My observations:

- Search is a huge improvement over what we got from Sphinx.
- The TOC architecture/organization/depth seems just right.
- Looks very polished and professional.
- This provides better maintainability.
- Let changes or revisions come in from the user community.

Overall, I'd say let's 🚢 and let our community find and report errors.

ecormany committed:
remove v2 docs from api project

ecormany committed:
restore tutorial protocol files

emilyburghardt committed:
docs(api, docs): add API 2.27 docs changes to mkdocs  (#20425)

# Overview

adding API 2.27 docs changes from these PRs: 

- [versioning updates for
2.27](https://github.com/Opentrons/opentrons/pull/20240/changes) ✅
- [dynamic pipetting
actions](https://github.com/Opentrons/opentrons/pull/20217) ✅
- [using capture_image in
8.8](https://github.com/Opentrons/opentrons/pull/20200) ✅
- [concurrent module
actions](https://github.com/Opentrons/opentrons/pull/20027) ✅

to the API docs in mkdocs project. 

TODO: 
- revisit versioning page and make sure all links redirect to the
headers I want them to. Need to check dynamic mix, and also can't get
the aspirate/dispense headers to link right in `liquids.md`. (both on
this page and in `robot-position.md`)
- properly linking `task` to the API reference? an outlier because I'm
linking a py class. in concurrent modules article and heater-shaker page
- concurrent module actions links in heater-shaker, temperature,
thermocycler pages not working right
- didn't add the edits to `new_protocol_api.rst` (CMAs PR), bc no
equivalent in mkdocs project?
- check that links to `task` build correctly in API ref: in
`module_contexts.py`, for `start_set_block_temperature()` and
`start_set_lid_temperature()`; in `protocol_context.py`, for
`wait_for_tasks()` and `create_timer()`
- proper links for `well` and `location` in `liquids.md` 

NEED TO `mkdocs serve -f
/Users/emilyburghardt/opentrons/docs/mkdocs.yml` for top level YML file
to build everything as accurately as possible

some link issues _might_ have to do with how I'm building my local
version? revisit this in 2026!


## Test Plan and Hands on Testing

sandbox (built locally)

## Changelog

List changes introduced by this PR considering future developers and the
end user. Give careful thought and clear documentation to breaking
changes.

## Review requests



## Risk assessment

definitely mid

ecormany committed:
hide 2.28 docstrings
